lib/apiservers/engine/backends/container.go

```diff
@@ -836,7 +836,40 @@ func (c *Container) ContainerInspect(name string, size bool, version version.Ver
 // ContainerLogs hooks up a container's stdout and stderr streams
 // configured with the given struct.
 func (c *Container) ContainerLogs(name string, config *backend.ContainerLogsConfig, started chan struct{}) error {
-	return fmt.Errorf("%s does not implement container.ContainerLogs", ProductName())
+	//FIXME:  Doug, remove following 3 lines once integration is added
+	if true {
+		return fmt.Errorf("%s does not implement container.ContainerLogs", ProductName())
+	}
+	defer trace.End(trace.Begin("ContainerAttach"))
+
+	// Look up the container name in the metadata cache to get long ID
+	vc := cache.ContainerCache().GetContainer(name)
+	if vc == nil {
+		return derr.NewRequestNotFoundError(fmt.Errorf("No such container: %s", name))
+	}
+	name = vc.ContainerID
+
+	tailLines, since, follow, err := c.validateContainerLogsConfig(vc, config)
+	if err != nil {
+		return err
+	}
+
+	// Outstream modification (from Docker's code) so the stream is streamed with the
+	// necessary headers that the CLI expects.  This is Docker's scheme.
+	wf := ioutils.NewWriteFlusher(config.OutStream)
+	defer wf.Close()
+	close(started)
+	wf.Flush()
+
+	var outStream io.Writer = wf
+	if !vc.Config.Tty {
+		outStream = stdcopy.NewStdWriter(outStream, stdcopy.Stdout)
+	}
+
+	// Make a call to our proxy to handle the remoting
+	err = c.containerProxy.StreamContainerLogs(name, outStream, started, config.Timestamps, follow, since, tailLines)
```

Comment:
Two things here that I've been noticing in the container code that are worth mentioning:

1. When we use the cache to convert from a container name -> container ID before making a backend call, we lose the name and are thus unable to present the error to the client in the same way docker does (displaying the name instead of the ID if that's what was supplied by the CLI). This especially obvious in cases where a containerVM has been removed out-of-band (using the GUI, for example). I've been storing off the `name` param in its original state and using that when returning `Not Found` type errors to the cli.

2. If the `err` we get back here happens to some type of container `NotFound` error (again, caused by some out-of-band container removal), we should delete the container from the container cache so later calls to `docker create --name mycontainer` won't return with `Already exists` type errors.

Don't worry about it here, I have a task open right now to address this, but I thought it was worth bringing to your attention.